src/report/report.c

```
@@ -0,0 +1,227 @@
+        if (r < 0)
+                return log_debug_errno(r, "Failed to issue io.systemd.Metrics.List call: %m");
+
+        return 0;
+}
+
+static void sd_varlink_unref_many(sd_varlink **array, size_t n) {
+        FOREACH_ARRAY(v, array, n)
+                sd_varlink_unref(*v);
+
+
+        free(array);
+}
+
+static int timeout_callback(sd_event_source *es, uint64_t usec, void *userdata) {
+        return sd_event_exit(ASSERT_PTR(sd_event_source_get_event(es)), EXIT_FAILURE);
+}
+
+static int metrics_start_query(void) {
+        _cleanup_closedir_ DIR *d = NULL;
+        _cleanup_(sd_event_unrefp) sd_event *event = NULL;
+        int r, c;
+
+        d = opendir(RUN_SYSTEMD_METRICS_PATH);
+        if (!d) {
+                if (errno == ENOENT)
+                        return -ESRCH;
+
+                return -errno;
+        }
+
+        r = sd_event_default(&event);
+        if (r < 0)
+                return log_error_errno(r, "Failed to get event loop: %m");
+
+        r = sd_event_set_signal_exit(event, true);
+        if (r < 0)
+                return log_error_errno(r, "Failed to enable exit on SIGINT/SIGTERM: %m");
+
+        sd_varlink *varlinks[MAX_CONCURRENT_METRICS_SOCKETS] = {};
+        size_t n_varlinks = MAX_CONCURRENT_METRICS_SOCKETS;
+        Context context = {};
+        int i = 0;
+
+        CLEANUP_ARRAY(varlinks, n_varlinks, sd_varlink_unref_many);
+
+        FOREACH_DIRENT(de, d, return -errno) {
+                _cleanup_free_ char *p = NULL;
+
+                if (de->d_type != DT_SOCK && de->d_type != DT_UNKNOWN)
+                        continue;
+
+                p = path_join(RUN_SYSTEMD_METRICS_PATH, de->d_name);
+                if (!p)
+                        return -ENOMEM;
+
+                context.n_ref++;
+
+                r = metrics_call(p, event, varlinks[i], &context);
+                if (r < 0) {
+                        context.n_ref--;
+                        continue;
```

please do not silently ignore errors. systemd does a special patter when ignoring errors:

```
log_at_suitable_level(r, "Failed to ...., ignoring");
```

---

added `log_error_errno()`